crates/net/eth-wire/src/types/broadcast.rs

```diff
@@ -529,103 +516,156 @@ impl HandleAnnouncement for NewPooledTransactionHashes66 {
         self.0.len()
     }
 
-    fn retain_by_hash(&mut self, mut f: impl FnMut(&TxHash) -> bool) -> Self {
-        let mut indices_to_remove = vec![];
-        for (i, hash) in self.0.iter().enumerate() {
-            if !f(hash) {
-                indices_to_remove.push(i);
-            }
-        }
+    fn dedup(self) -> PartiallyValidData<Self::Value> {
+        let Self(hashes) = self;
 
-        let mut removed_hashes = Vec::with_capacity(indices_to_remove.len());
+        let mut deduped_data = HashMap::with_capacity(hashes.len());
 
-        for index in indices_to_remove.into_iter().rev() {
-            let hash = self.0.remove(index);
-            removed_hashes.push(hash);
+        let noop_value: TxMetadata = None;
+
+        for hash in hashes.into_iter().rev() {
+            deduped_data.insert(hash, noop_value);
         }
 
-        Self(removed_hashes)
+        PartiallyValidData::from_raw_data_eth66(deduped_data)
     }
+}
 
-    fn msg_version(&self) -> EthVersion {
-        EthVersion::Eth66
-    }
+/// Interface for handling announcement data in filters in the transaction manager and transaction
+/// pool. Note: this trait may disappear when distinction between eth66 and eth68 hashes is more
+/// clearly defined, see <https://github.com/paradigmxyz/reth/issues/6148>.
+pub trait HandlePayload {
+    /// The announcement contains no entries.
+    fn is_empty(&self) -> bool;
+
+    /// Returns the number of entries.
+    fn len(&self) -> usize;
+
+    /// Retain only entries for which the hash in the entry satisfies a given predicate, return
+    /// the rest.
+    fn retain_by_hash(&mut self, f: impl FnMut(&TxHash) -> bool) -> Self;
```

Comment:
when are we calling this?

Reply:
in pool and in tx manager, upon filtering based on various predicates. should be indexed in GitHub so if you click on the trait method definition, you should get links to where it's called.

Reply:
hmm, I don't like how this always reallocates.

do we even need the value from the response?
if we only need the hash then we can use retain and return the hashes as vec

Reply:
I think ideally we want `extract_if` which is currently nightly only

but hashbrown has it on stable

https://docs.rs/hashbrown/latest/hashbrown/struct.HashMap.html#method.extract_if

Reply:
I think we don't want to use `ahash` which is a dep of hash brown  https://github.com/paradigmxyz/reth/pull/6681#issuecomment-1953434717. due to @Rjected 's cryptography background, which I don't have personally, I will second his point of view on `ahash`.